### PR TITLE
feat: Oracle mocker for nargo test

### PR DESCRIPTION
**Pull request by sirasistant**
# Description
Adds an oracle mocker feature to be able to do noir unit tests for code that uses oracles.

## Problem\*

Noir test part for https://github.com/noir-lang/noir/issues/1052 and https://github.com/AztecProtocol/aztec-packages/issues/1642

## Documentation

- [ ] This PR requires documentation updates when merged.

  

  - [ ] I will submit a noir-lang/docs PR.

  

  - [ ] I will request for and support Dev Rel's help in documenting this PR.

  
  

## Additional Context

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

**Comment by sirasistant** [comment]
Should not be breaking after split into this and https://github.com/noir-lang/noir/pull/2935

File: tooling/nargo/src/ops/foreign_calls.rs [hunk 1]
```diff
@@ -76,10 +126,92 @@ impl ForeignCall {
...
+            Some(ForeignCall::SetMockTimes) => {
+                let (id, params) = Self::extract_mock_id(&foreign_call.inputs)?;
+                let times = params[0]
+                    .unwrap_value()
+                    .to_field()
+                    .try_to_u64()
+                    .expect("Invalid bit size of times");
+
+                self.find_mock_by_id(id)
+                    .unwrap_or_else(|| panic!("Unknown mock id {}", id))
```
Comment (vezenovm) [inline tooling/nargo/src/ops/foreign_calls.rs:163]: These panics could be turned into errors, but this should be rare as it is an internal compiler error so this is non-blocked if we need to get this in. `ForeignCallError` was moved into `noirc_printable_type` so we probably could do a general error refactor in a separate PR.

File: tooling/nargo/src/ops/foreign_calls.rs [hunk 2]
```diff
@@ -76,10 +126,92 @@ impl ForeignCall {
...
+                    .times_left = Some(times);
+
+                Ok(ForeignCallResult { values: vec![] })
+            }
+            Some(ForeignCall::ClearMock) => {
+                let (id, _) = Self::extract_mock_id(&foreign_call.inputs)?;
+                self.mocked_responses.retain(|response| response.id != id);
+                Ok(ForeignCallResult { values: vec![] })
+            }
+            None => {
```
Comment (vezenovm) [inline tooling/nargo/src/ops/foreign_calls.rs:173]: So we are going to be assuming from this point on that any user-defined oracle is to be used as a mock oracle by `nargo`? I guess this is fine as any new oracles we would like to add will be need to be implemented in the foreign call executor with their own match case.

File: tooling/nargo/src/ops/foreign_calls.rs [hunk 3]
```diff
@@ -36,16 +45,57 @@ impl ForeignCall {
...
     }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+struct MockedCall {
+    id: usize,
+    name: String,
+    params: Option<Vec<ForeignCallParam>>,
+    result: ForeignCallResult,
+    times_left: Option<u64>,
```
Comment (vezenovm) [inline tooling/nargo/src/ops/foreign_calls.rs:64]: Could we document a bit more the purpose of `times_left`, and then also why we have a `last_mock_id` inside of the ForeignCallExecutor

File: tooling/nargo/src/ops/foreign_calls.rs [hunk 4]
```diff
@@ -76,10 +126,92 @@ impl ForeignCall {
                     ],
                 })
             }
-            None => panic!("unexpected foreign call {foreign_call_name:?}"),
+            Some(ForeignCall::CreateMock) => {
+                let mock_oracle_name = Self::parse_string(&foreign_call.inputs[0]);
+                assert!(ForeignCall::lookup(&mock_oracle_name).is_none());
```
Comment (vezenovm) [inline tooling/nargo/src/ops/foreign_calls.rs:131]: This could be its own error as well, but like the other comment I am fine leaving this for another PR.

**Review by vezenovm — COMMENTED** [review]
LGTM just a few small comments

File: tooling/nargo/src/ops/foreign_calls.rs (hunk 1 above)
Comment (sirasistant) [reply tooling/nargo/src/ops/foreign_calls.rs:163]: Yeah it felt pretty weird to add these mocking specific errors in noirc_printable_type

File: tooling/nargo/src/ops/foreign_calls.rs (hunk 2 above)
Comment (sirasistant) [reply tooling/nargo/src/ops/foreign_calls.rs:173]: This is for testing only, If we implement a solution for oracles in nargo execute, our solution should contact the custom resolver (inter-process communication or whatever solution we choose) in the case there are no mocks active :thinking:

Comment (vezenovm) [reply tooling/nargo/src/ops/foreign_calls.rs:181]: Yeah a full solution would require that. It looks like this is good to do as `nargo execute` is only used for testing and println, and if someone tries to define their own oracle to not used in a mock they will get an unknown foreing call error.

**Review by vezenovm — DISMISSED** [review]
We can update the errors in a separate PR
